state/upgrade/types.go

```diff
@@ -67,3 +67,21 @@ func (a OldAddress27) Upgrade(lookup network.SpaceInfos) (OldAddress27, error) {
 	}
 	return a, nil
 }
+
+// OldPortsDoc28 represents a ports document prior to the 2.9 schema changes.
+type OldPortsDoc28 struct {
+	DocID     string              `bson:"_id"`
+	ModelUUID string              `bson:"model-uuid"`
+	MachineID string              `bson:"machine-id"`
+	SubnetID  string              `bson:"subnet-id"`
+	Ports     []OldPortRangeDoc28 `bson:"ports"`
+	TxnRevno  int64               `bson:"txn-revno"`
+}
+
+// OldPortsDoc28 represents a port range entry document prior to the 2.9 schema changes.
+type OldPortRangeDoc28 struct {
+	UnitName string
+	FromPort int
+	ToPort   int
+	Protocol string
```

Comment:
Should tag these

Reply:
Should we always put bson tags in general? The original type in state before my changes did not (hence the odd lowercased key names in the `$set` queries). I retained this for the new type but should probably tag those as well to get better looking keys.

Reply:
@wallyworld It looks like the same functionality for open/close/conflicts is tested multiple times (I counted at least 3 different suites in various places that test the same thing over and over) and I believe the rollup test to be adequate (I mean, the rollup logic is pretty trivial). Do you have any particular use case in mind that I may have missed?

That being said, I do have a card for converting the QA step from all these related PRs into functional tests (once all bits land) so I think we should be able to cover more things that way.

Reply:
It was nothing in particular - just a (not always followed) convention that we should tag struct attributes for serialisation, even if the tags would match the default serialisation names. It means there's absolutely no ambiguity. But since this is just a type used for an upgrade step, it doesn't really matter so mych.